### DISCUSSION TITLE
Only render changed files

### DISCUSSION THREAD

**Discussion by bryanwhiting** [quarto-dev/quarto-cli#3674] [Q&A] [answered]
On of the benefits of Hugo is how fast it is. I don't like how slow quarto is when rendering (taking several seconds to build + refresh). I wrote this script to scan for files that have changed locally compared to remote and only render those files. It takes the guesswork out of comparing changes.

```
library(dplyr)
library(stringr)
files <- system("git diff main origin/main --name-only -- . :^docs", intern=T) %>% 
    tibble(files=.) %>% 
    filter(str_detect(files, 'md$')) %>% 
    pull(files)

if(length(files) > 0){
  cat(
    paste0('Rendering uncommitted *.md files:\n\t', 
           paste0(files, collapse ="\n\t")
    ), 
  "\n")
  for(f in files){
    cmd <- paste0("quarto render --quiet ", f)
    print(cmd)
    system(cmd)
  }
}
```

**Comment by mcanouil** [comment]
Did you try `execute.freeze`?
See <https://quarto.org/docs/projects/code-execution.html#freeze>

**Comment by bryanwhiting** [comment]
Yes, didn't "skip" the file like I had hoped when running a blanket "quarto
render"

**Reply by bryanwhiting** [reply]
I have execute.freeze set in both about.qmd and `posts/...file.qmd` and it still spends several seconds rendering all these files.

```
---
title: "About"
image: img/profile.jpg
execute:
    freeze: true
```

**Reply by dragonstyle** [reply]
This is expected as `freeze` is really designed to freeze execution of the code in the QMD (versus the actual rendering of the document). Huge is definitely faster than Quarto and we currently have erred on the side of caution and made render always render the full site to ensure that the output always reflects the current saved input state. 

This would be a good area for us to improve over time...

**Reply by bryanwhiting** [reply]
yes, i assumed this was expected. What's the rationale for rendering files that haven't changed? I'd image a `quarto render --all` or `--force` would be an expected behavior to render all posts and `quarto render` to just do those that have changed.

**Reply by dragonstyle** [reply]
I think it's less about a principled position and more that we don't currently have a reliable enough render state to know that the input file hasn't changed. As a result, we always render. I think it would be something we'd like to get better / faster at!

**Reply by bryanwhiting** [reply]
ahh, makes sense. Thanks for taking this into consideration! love the work you're doing. I considered moving back to hugo because the rendering was slow (I miss Live Reload), but Quarto has so many rich features that this is one I can do without for now :)  Thanks for all you're building!

**Comment by mcanouil** [comment]
The freeze option is a project option: "Note that freeze controls whether execution occurs during global project renders."

This might also be related to #3661

**Reply by bryanwhiting** [reply]
yes, that speaks to it

**Comment by cscheid** [comment, marked as answer]
That's a neat script. Making quarto faster is definitely on our radar and there's some incoming improvements in 1.3.

With regards to re-rendering, we're currently being conservative on `quarto render` because of possible dependencies. As a simple example, listings can use content from other files, and so a simple `diff`-based approach won't always work.

If you're working interactively on a large quarto project, then `quarto preview` can offer a better experience than `quarto render` everything all the time: the preview command will render everything once, but then will only refresh the changed files.

**Reply by bryanwhiting** [reply]
Thanks, I haven't been using `quarto preview` when I now realize I should be. The problem is that quarto preview wasn't working inside my docker container and so I went with this approach. Maybe I just need to figure out why it's not working in my docker image.

**Reply by bryanwhiting** [reply]
I tried 

```rstudio@d1e3602ac4a0:~/blog$ quarto preview
Preparing to preview

Watching files for changes
http://localhost:3000/
```
and I have the port mounted: 

```
docker run \
	--detach \
	--publish 80:8787 \
	--publish 3000:3000 \
	--env DISABLE_AUTH=true \
	--name renv \
	--volume $GH/blog:$ROOT/blog \
	rocker/tidyverse
```

but when I go to `localhost:3000`  I just get:
```
This page isn’t workinglocalhost didn’t send any data.
ERR_EMPTY_RESPONSE
```

**Reply by mcanouil** [reply]
See <https://github.com/quarto-dev/quarto-cli/discussions/2083>

**Reply by josherrickson** [reply]
> If you're working interactively on a large quarto project, then quarto preview can offer a better experience than quarto render everything all the time: the preview command will render everything once, but then will only refresh the changed files.

Is there a technical reason that either `quarto render` can't be given an option to operate like `quarto preview` (just in terms of only re-rendering changed files), or that `quarto preview` can't be given an option to re-render then exit without keeping running? I've tried adding the the various `--no-` options, and `--timeout 0`/`--timeout 1` but can't seem to make it happen. The use case is to replace `quarto render` to just build the site, without having the Ctrl-C afterwards.

**Reply by ThSGM** [reply]
> That's a neat script. Making quarto faster is definitely on our radar and there's some incoming improvements in 1.3.
> 
> With regards to re-rendering, we're currently being conservative on `quarto render` because of possible dependencies. As a simple example, listings can use content from other files, and so a simple `diff`-based approach won't always work.
> 
> If you're working interactively on a large quarto project, then `quarto preview` can offer a better experience than `quarto render` everything all the time: the preview command will render everything once, but then will only refresh the changed files.

It's been a couple of years since this was raised, and despite my searching around these parts, I'm not sure this was ever addressed.

I have a Quarto book that includes about 30 chapters, say. This is rendered into PDF so needs to run through xelatex. In order to view the PDF file, I need to re-render the complete document each time!

**Comment by royfrancis** [comment]
This is extremely useful in a project with large number of files and collaborators.

In addition to freezing code execution, there must be an option to freeze rendering of files in a project. Something like

```
render:
    freeze: true
```

Another idea is to have it linked to format to allow selective freezing of specific formats in a project. Something like

```
format:
  html:
    freeze: true
```

I totally understand the concern "we don't currently have a reliable enough render state to know that the input file hasn't changed". But, leave that to the user to decide whether they want to use this option and take the risk.

**Reply by allefeld** [reply]
I second this. Make it a YAML option or a command line option, e.g. `--only-changed`.

Use case: I have a Quarto website which contains a large number of source files. I usually only change one of them at a time, but all get rendered which takes a lot of time. That's unnecessary because there are no dependencies between them.

`quarto preview` may serve as a hack in the meantime, but it doesn't exactly provide the needed functionality, because it is watching for file changes.

**Reply by allefeld** [reply]
Plus, this is also nicer for VCS. This way, all html files are updated in every commit.

**Comment by mcanouil** [comment]
To note, the following "epic" issue should help with this:
- #6518

**Reply by bryanwhiting** [reply]
thanks!

**Comment by HaykTarkhanyan** [comment]
@bryanwhiting Thank you for the script. It saves a lot of time for me.

**Comment by bryanwhiting** [comment]
Glad it's still helping out all these years later!

Thanks,

Bryan

**Comment by prosoitos** [comment]
I also hope that this will be implemented in the future.

In my case, I have a website with a growing number of reveal.js presentations in it.

Freezing works perfectly for the website pages, but it has no effect for the reveal.js slides decks. This causes:

- A very slow render of the site (the 400 or so pages of the site go rather fast thanks to freezing, but the reveal.js presentations each take a very long time since they are re-rendered every single time).
- The presentations keep breaking when Quarto, Mermaids, etc. get updated since I cannot freeze them at a point in time and leave them alone.

Thanks for working on this!

Notes:

- `freeze: true` doesn't have any more effect than `freeze: auto` for the reveal.js slides decks inside the project.

- I know that I am not the only one facing this issue with reveal.js presentations in a website (it is a classic setup for teaching) and I know that others go around the current limitation by rendering their sites with a Makefile (to automate the rendering of only those files that changed) instead of doing a simple render of the whole site from the root of the project. I think that I will start doing it too.

Now, this workaround might cause some of the issues that you are trying to avoid by re-rendering some of the unmodified files.

**Comment by venpopov** [comment]
It has now been several years since this issue was raised not only here, but in various other blog posts and discussion forums. Personally I love Quarto and use it for all my projects, but the rendering speed remains the biggest annoyance I have. On my blog, for example, I have about a couple of static pages and a dozen blog posts, each relatively short, and rendering the blog still takes 20 seconds with everything frozen and cached:

```
quarto render
[ 1/16] CV/index.qmd
[ 2/16] index.qmd
[ 3/16] posts.qmd
[ 4/16] posts/2024/can-i-reproduce-my-own-old-research/index.qmd
[ 5/16] posts/2024/git-local-ignore/index.qmd
[ 6/16] posts/2024/introducing-bmm/index.qmd
[ 7/16] posts/2024/r-multiple-package-versions/index.qmd
[ 8/16] posts/2024/reduce-friction-for-creating-quarto-blog-posts/index.qmd
[ 9/16] posts/2024/reproducibility-is-hard/index.qmd
[10/16] posts/2024/reproducible-r-projects/index.qmd
[11/16] posts/2024/taking-a-stand-on-open-peer-review/index.qmd
[12/16] posts/2025/arto-pollo/index.qmd
[13/16] posts/2025/open-science-needs-reliable-infrastructure/index.qmd
[14/16] posts/2025/r-love-you-r-hate-you/index.qmd
[15/16] posts/2026/package-management-and-reproducibility-are-a-nightmare/index.qmd
[16/16] publications.qmd
```

The plain qmd files are relatively quick, but any file with code, even if the code execution is frozen, still takes a surprising amount of time.

It is not a deal-breaker, but over time the annoyance and the waiting builds up

**Reply by cscheid** [reply]
We hear you very loudly, I promise! Two comments:

1. The current (very poor) performance characteristics of Quarto come from a large number of architectural decisions that are not easy to work within the current architecture of Quarto. It's just as annoying to us as it is to you (and some of our colleagues at Posit have to render Quarto projects with thousands of files. They know my email! They tell me it's terrible! They're right!)
2. With all that said, Stay Tuned. We'll have news in 2026.

**Reply by judgelord** [reply]
Very excited to hear this. One small note, just because I don't see this aspect has yet been raised: 

Websites that include PDFs and DOCXs as "Other Formats" might be a special case: an option to simply not re-render secondary PDF and DOCX formats may be easier to implement and have a high impact on speed.

We may often want to re-render all of the HTML of a website (e.g., for style updates) but not want to regenerate the other formats of unmodified documents. 

Thanks for all your great work!

**Reply by prosoitos** [reply]
I think that it would be best if any document with any other format within a website weren't re-rendered by a site render. That would solve your situation with PDFs and DOCXs as well as mine (and that of many colleagues of mine) with reveal.js talks within the website.

**Reply by prosoitos** [reply]
So I don't see those as a special case. I think that a website render should render the pages of the website itself and any document with another format within the website project should be re-rendered explicitly and separately.

**Reply by judgelord** [reply]
Perhaps. I'd often prefer my slide decks not render. Other times, I want them all to inherit a new scss theme or JavaScript accessibility patch (but there is no reason to update all my PDFs when messing with HTML styling). 

The reverse may also occur if I'm updating a PDF template for the PDF versions of all documents, but I suspect this is much less common and typically much less of a problem in terms of wasted time.

Hopefully, the fix has options for what formats not to render if unchanged. 

I just suspected that non-indexed/non-HTML formats would be easier to carve off from the rest of the site architecture and thus easier to implement. (That's all I meant by a special case.)

**Reply by prosoitos** [reply]
Right. That's true. You convinced me. So re-rendering the reveal.js in a website might in the end still be the correct solution (while excluding PDFs and other non-HTML documents). So the solution for the reveal.js would be for the cached/frozen code to work (right now, there is no way to have the cache used and the code is re-executed in reveal.js slides, which is my true problem).
